bazel: upgrade to clang 19

rockwotj opened the pull request:
Also don't add the compiler name to the repo, it just makes it tedious
to update, now upgrading clang versions should only require changes to
MODULE.bazel

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

rockwotj committed:
bazel: upgrade to clang 19

Also don't add the compiler name to the repo, it just makes it tedious
to update, now upgrading clang versions should only require changes to
MODULE.bazel

rockwotj commented:
@clee can we update CI so that PRs like this run the full PR pipeline? Thanks!

rockwotj commented:
(in the meantime I will reformat the codebase to force CI to run)

rockwotj committed:
treewide: run clang-format-19

rockwotj committed:
tools: run clang-format in parallel

rockwotj committed:
lint: format seed-maven-cache

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5b3cdd6c-fdfa-43a5-809f-cb401c32bb43-->
<!--TEST RESULTS-->
<!--a7cd3290-8599-40ea-b153-c71ba32738f0-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65870">build#65870</a></summary>


| test_class                | test_method            | test_arguments                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:-----------------------|:-----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_chunks       |                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65870#0196c812-a06a-4884-9845-3486a65f15bd | FLAKY         | 20/21    | upstream reliability is '99.53703703703704'. current run reliability is '95.23809523809523'. drift is 4.29894 and the allowed drift is set to 50. The test should PASS |
| DatalakeTransactionTests  | test_with_transactions | {"cloud_storage_type": 1, "compaction": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65870#0196c812-a06a-4884-9845-3486a65f15bd | FLAKY         | 20/21    | upstream reliability is '99.53703703703704'. current run reliability is '95.23809523809523'. drift is 4.29894 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66796">build#66796</a></summary>


| test_class                | test_method                                         | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:----------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_when_segment_size_smaller_than_chunk_size |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66796#0197392b-11f1-4a78-a5b1-e2285e0f8508 | FLAKY         | 20/21    | upstream reliability is '96.93877551020408'. current run reliability is '95.23809523809523'. drift is 1.70068 and the allowed drift is set to 50. The test should PASS |
| PartitionBalancerTest     | test_fuzz_admin_ops                                 |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66796#0197392b-11f1-4a78-a5b1-e2285e0f8508 | FLAKY         | 19/21    | upstream reliability is '99.51923076923077'. current run reliability is '90.47619047619048'. drift is 9.04304 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations                                | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66796#0197392b-11f3-4c1b-b115-fcab847255eb | FLAKY         | 19/21    | upstream reliability is '98.51851851851852'. current run reliability is '90.47619047619048'. drift is 8.04233 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj commented:
/microbench

rockwotj committed:
thirdparty: update jsoncons

rockwotj committed:
toolchain: upgrade to clang 20

rockwotj committed:
storage/tests: remove deprecated syntax from literal operator

The space is not allowed in clang 20

rockwotj committed:
controller_api: remove unused field

Caught by clang 20

rockwotj committed:
config_manager: remove unused field

Caught by clang 20

rockwotj committed:
datalake: remove unused fields

Thanks clang 20!

rockwotj committed:
tx_topic_manager: remove unused field

Thanks clang 20

rockwotj committed:
kafka/client: fix clang 20 error

> chosen constructor is explicit in copy-initialization

rockwotj committed:
iceberg: remove unused field

Thanks clang 20!

rockwotj committed:
iceberg: remove unused fields in merge append action

rockwotj committed:
persisted_stm: use unused field

_log was unused, we can use it in the same way it's used for the file
based stm snapshots

rockwotj committed:
members_frontend: remove unused field

rockwotj committed:
config_frontend: remove unused member

rockwotj committed:
topic_updates_dispatcher: remove unused member

rockwotj committed:
members_backend: remove unused member

rockwotj committed:
cluster/partition: remove unused field

rockwotj committed:
health_manager: remove unused member

rockwotj committed:
cluster_recovery_manager: remove unused field

rockwotj committed:
tm_stm: remove feature_table

rockwotj committed:
ephemeral_credential_frontend: remove unused member

rockwotj committed:
thirdparty: revert jsoncons upgrade

Instead just apply the patch to make the code build for clang 20 without
errors. There seems to be some kind of breaking change between what we
are using and the latest verison. Instead of digging into that we'll
just live with the patch for now.

rockwotj commented:
Blocked by https://github.com/llvm/llvm-project/issues/141858

We can't use clang-19 due to clang-tidy crashes.

redpanda-gandalf commented:
@rockwotj Do you want to move it back to draft?

dotnwat commented:
> Blocked by [llvm/llvm-project#141858](https://github.com/llvm/llvm-project/issues/141858)
> 
> We can't use clang-19 due to clang-tidy crashes.

could we run a different version of clang-tidy?

```
build:clang-tidy --@bazel_clang_tidy//:clang_tidy_config=//:clang_tidy_config
```

rockwotj committed:
bazel: upgrade to clang-19

But hold clang-tidy back due to crashes

rockwotj committed:
treewide: format the codebase with clang-19

rockwotj commented:
> > Blocked by [llvm/llvm-project#141858](https://github.com/llvm/llvm-project/issues/141858)
> > We can't use clang-19 due to clang-tidy crashes.
> 
> could we run a different version of clang-tidy?
> 
> ```
> build:clang-tidy --@bazel_clang_tidy//:clang_tidy_config=//:clang_tidy_config
> ```

This is a good idea. Held clang-tidy back to 18, moved everything else to 19. Hopefully we can upgrade to 20 and get back to same version everywhere.

rockwotj committed:
lint: reformat shell scripts

rockwotj commented:
Ready for review

dotnwat committed:
bazel: apply bazel mod tidy

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

rockwotj commented:
Thanks Noah for the assist here!
